go-controller/pkg/node/base_node_network_controller_dpu.go

```diff
@@ -14,139 +13,214 @@ import (
+		return true
+	}
+	if oldDPUCD.PfId != newDPUCD.PfId ||
+		oldDPUCD.VfId != newDPUCD.VfId || oldDPUCD.SandboxId != newDPUCD.SandboxId {
+		return true
+	}
+	return false
+}
+
 // watchPodsDPU watch updates for pod dpu annotations
 func (bnnc *BaseNodeNetworkController) watchPodsDPU() error {
-	var retryPods sync.Map
-	// servedPods tracks the pods that got a VF
-	var servedPods sync.Map
-
 	clientSet := cni.NewClientSet(bnnc.client, corev1listers.NewPodLister(bnnc.watchFactory.LocalPodInformer().GetIndexer()))
 
-	// Support default network for now
-	nadName := types.DefaultNetworkName
 	netName := types.DefaultNetworkName
 	_, err := bnnc.watchFactory.AddPodHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			pod := obj.(*kapi.Pod)
-			klog.Infof("Add for Pod: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
+			klog.V(5).Infof("Add for Pod: %s/%s for network %s", pod.Namespace, pod.Name, netName)
 			if util.PodWantsHostNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
 				return
 			}
-			if util.PodScheduled(pod) {
-				// Is this pod created on same node as the DPU
-				if bnnc.name != pod.Spec.NodeName {
-					return
-				}
+			// add all the Pod's NAD into Pod's podNadCache
+			// networkMap is network selection element map, key is the NAD name. For default network,
+			// NAD name is DefaultNetworkName.
+			networkMap := make(map[string]*nettypes.NetworkSelectionElement)
+			// Support default network for now
+			networkMap[types.DefaultNetworkName] = nil
+			bnnc.podNadCache.Store(pod.UID, networkMap)
 
-				vfRepName, err := bnnc.getVfRepName(pod)
-				if err != nil {
-					klog.Infof("Failed to get rep name, %s. retrying", err)
-					retryPods.Store(pod.UID, true)
-					return
-				}
-				isOvnUpEnabled := atomic.LoadInt32(&bnnc.atomicOvnUpEnabled) > 0
-				podInterfaceInfo, err := cni.PodAnnotation2PodInfo(pod.Annotations, nil, isOvnUpEnabled, string(pod.UID),
-					"", nadName, netName, config.Default.MTU)
-				if err != nil {
-					retryPods.Store(pod.UID, true)
-					return
+			// initialize serverCache to be empty
+			isOvnUpEnabled := atomic.LoadInt32(&bnnc.atomicOvnUpEnabled) > 0
+			servedCache := map[string]*util.DPUConnectionDetails{}
+			for nadName := range networkMap {
+				dpuCD := bnnc.podReadyToAddDPU(pod, nadName)
+				if dpuCD != nil {
+					err := bnnc.addDPUPod4Nad(pod, dpuCD, isOvnUpEnabled, netName, nadName, clientSet)
+					if err == nil {
+						servedCache[nadName] = dpuCD
+					}
 				}
-				err = bnnc.addRepPort(pod, vfRepName, podInterfaceInfo, clientSet)
-				if err != nil {
-					klog.Infof("Failed to add rep port, %s. retrying", err)
-					retryPods.Store(pod.UID, true)
-				} else {
-					servedPods.Store(pod.UID, true)
-				}
-			} else {
-				// Handle unscheduled pods later in UpdateFunc
-				retryPods.Store(pod.UID, true)
-				return
 			}
+			bnnc.servedPods.Store(pod.UID, servedCache)
```

Comment:
It seems that the 2 maps `bnnc.podNadCache` and `bnnc.servedPods` can be combined. They should share the same key which is the NAD name. We can have a struct that stores both data structures.

Reply:
servedCache per Pod only records the NADs that have been successfully configured, while  podNadCache per Pod records all NADs that needs to be configured.

Reply:
Wouldn't it be more readable if there was some state (maybe something along the lines of SUCCESS_CONFIG, NEEDS_CONFIG) and we only maintain 1 map (which only requires 1 existence check). The state variable will be maintained in the map.

Reply:
ok, I will try to merge these two maps.

Reply:
Thanks! It looks more readable now.